packages/cardano-services/test/Asset/CardanoTokenRegistry.test.ts

```diff
@@ -0,0 +1,158 @@
+    resolver = resolve;
+    rejecter = reject;
+  });
+
+  server.listen(port, 'localhost');
+  server.on('error', rejecter);
+
+  return {
+    closeMock: () => {
+      server.close((error) => (error ? rejecter(error) : resolver()));
+      return closePromise;
+    },
+    metadataServerUri: `http://localhost:${port}`
+  };
+};
+
+describe('CardanoTokenRegistry', () => {
+  const nonvalidAssetId = Cardano.AssetId('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef');
+  const validAssetId = Cardano.AssetId('f43a62fdc3965df486de8a0d32fe800963589c41b38946602a0dc53541474958');
+
+  describe('return value', () => {
+    const tokenRegistry = new CardanoTokenRegistry({ logger: dummyLogger });
+
+    afterAll(() => tokenRegistry.shutdown());
+
+    it('returns null for nonexisting AssetId', async () => {
+      expect(await tokenRegistry.getTokenMetadata([nonvalidAssetId])).toEqual([null]);
+    });
+
+    it('returns metadata when subject exists', async () => {
+      const [metadata] = await tokenRegistry.getTokenMetadata([validAssetId]);
+
+      expect(metadata).not.toBeNull();
+
+      const { icon, ...rest } = metadata!;
+      const result: Asset.TokenMetadata = {
+        decimals: 8,
+        // eslint-disable-next-line max-len
+        desc: "SingularityNET lets anyone - create, share, and monetize AI services at scale. SingularityNET is the world's first decentralized AI network",
+        name: 'SingularityNet AGIX Token',
+        ticker: 'AGIX',
+        url: 'https://singularitynet.io/'
+      };
+
+      // We do not check the entire icon value as it is a long long string
+      expect(icon).toBeDefined();
+      expect(rest).toEqual(result);
+    });
+
+    it('correctly returns null or metadata for request with good and bad assetIds', async () => {
+      const firstResult = await tokenRegistry.getTokenMetadata([nonvalidAssetId, validAssetId]);
+      const secondResult = await tokenRegistry.getTokenMetadata([validAssetId, nonvalidAssetId]);
+
+      expect(firstResult[0]).toBeNull();
+      expect(firstResult[1]).not.toBeNull();
+      expect(secondResult[0]).not.toBeNull();
+      expect(secondResult[1]).toBeNull();
+    });
+  });
+
+  it('metadata are cached', async () => {
+    const gotValues: unknown[] = [];
+
+    class TestInMemoryCache extends InMemoryCache {
+      public getVal<T>(key: Key): T | undefined {
+        const value = super.getVal<T>(key);
+
+        gotValues.push(value);
+
+        return value;
+      }
+    }
+
+    const tokenRegistry = new CardanoTokenRegistry({ cache: new TestInMemoryCache(1), logger: dummyLogger });
+    const [firstResult] = await tokenRegistry.getTokenMetadata([validAssetId]);
+    const [secondResult] = await tokenRegistry.getTokenMetadata([validAssetId]);
+
+    expect(gotValues[0]).toBeUndefined();
+    expect(gotValues[1]).toEqual(firstResult);
+    expect(firstResult).toEqual(secondResult);
+
+    tokenRegistry.shutdown();
```

Comment:
`shutdown`  will not be executed in case of failed assertion

Reply:
Resolved in d7831d36cf8126eab8f875065a6aa581259e2141

could you please review this again @iadmytro ? Thank you